src/v/serde/test/serde_test.cc

```diff
@@ -693,6 +715,7 @@ struct old_cs
     bool operator==(const old_cs&) const = default;
 
     std::vector<test_msg1_new_manual> data_;
+    auto serde_fields() { return std::tie(data_); }
```

Comment:
the template name old_cs and old_no_cs in CRTP look wrong (unless its on purpose for the test)

Reply:
sigh. these are all over the tests. no clue how to guess whether it's intentional.

Reply:
we can ask @felixguendling and see if he has a memory

Reply:
I think this is a copy & paste mistake I made. Back then, the type was not used by the `envelope` class at all, so maybe this is why it was overlooked.

Reply:
Cool, thanks @felixguendling. I've made this mistake a handful of times, and I only catch it by visual inspection. Not sure whether it's testable in a general way in practice. Maybe as a clang-tidy check or something.

Reply:
I tried something, but due to the cyclic nature, it didn't really work:


```cpp
#include <type_traits>

template <typename T>
struct p {
    using value_t = T;

    // does not compile
    // definition of 'p<c>' is not complete until the closing '}'
    // static_assert(std::is_assignable_v<p<T>, T>);
};

struct c : p<c> {};
static_assert(std::is_assignable_v<p<c>, c>);  // correct, no problem

struct w : p<c> {};
static_assert(std::is_assignable_v<p<w>, w>);  // assertion fails
```

https://gcc.godbolt.org/z/baTocTqG9

So, yes - probably only a clang-tidy check or something similar would help.

Reply:
With some manual overhead, you can check every envelope type after it's defined:

```cpp
#include <type_traits>

template <typename T>
struct p {
    using value_t = T;
};

template <typename T>
concept Good = std::is_assignable_v<T, typename T::value_t>;

struct c : p<c> {};
static_assert(Good<c>);

struct w : p<c> {};
static_assert(Good<w>);  // fails
```

Reply:
>cyclic nature

Yeah, exactly. Manual check lgtm; I could envision adding a debug build step where we scrape headers for envelopes then  generate and collect static asserts in a separate compilation unit, but that's probably rather expensive to compile. Besides which it wouldn't cover this case, where the envelope is isolated in a `.cc` file.